src/AutoRest.CSharp/Common/Input/Source/ProtocolCompilationInput.cs

```diff
@@ -0,0 +1,99 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+using AutoRest.CSharp.AutoRest.Plugins;
+using AutoRest.CSharp.Generation.Types;
+using Microsoft.CodeAnalysis;
+
+namespace AutoRest.CSharp.Input.Source
+{
+    public class ProtocolCompilationInput : CompilationInput
+    {
+        private List<IMethodSymbol>? _methodSet;
+
+        public static async Task<CompilationInput?> TryCreate(string? existingProjectFolder)
+        {
+            return Configuration.ExistingProjectFolder != null ?
+                new ProtocolCompilationInput(await GeneratedCodeWorkspace.CreateExistingCodeProject(Configuration.ExistingProjectFolder).GetCompilationAsync()) : null;
+        }
+
+        private ProtocolCompilationInput(Compilation compilation)
+            : base(compilation) { }
+
+        private protected override void FilterSymbols()
+        {
+            _methodSet = new List<IMethodSymbol>();
+            foreach (IModuleSymbol module in _compilation.Assembly.Modules)
+            {
+                foreach (var type in SourceInputHelper.GetSymbols(module.GlobalNamespace))
+                {
+                    if (type is INamedTypeSymbol typeSymbol && IsClient(typeSymbol))
+                    {
+                        foreach (var member in typeSymbol.GetMembers())
+                        {
+                            if (member is IMethodSymbol methodSymbol && IsProtocolMethod(methodSymbol))
+                            {
+                                _methodSet.Add(methodSymbol);
+                            }
+                        }
+                    }
+                }
+            }
+        }
+
+        internal override IMethodSymbol? FindMethod(string namespaceName, string typeName, string methodName, IEnumerable<CSharpType> parameters)
+        {
+            if (_methodSet == null)
+            {
+                FilterSymbols();
+            }
```

Comment:
we could directly use the property `MethodSet` instead in this method, not to worry about "whether this property is initialized or not" because that is covered by the property.

Reply:
You mean we add a property `MethodSet` with backing field `_methodSet`? What is the benefit?